### PR TITLE
util: Cgroup monitoring and alerting

### LINKED ISSUE #50467: Cgroup monitoring and alerting
## Feature Request

**Is your feature request related to a problem? Please describe:**

There isn't yet a mechanism to monitor the cgroup cpu/memory limit and do corresponding alerting.

**Describe the feature you'd like:**

Add a monitor to watch cgroup cpu/memory limit setting changes and update corresponding metrics. Display the limits on related Grafana graphs. Trigger alerting when cpu/memory usage exceeds a threshold relative to the limit.

**Describe alternatives you've considered:**

None.

**Teachability, Documentation, Adoption, Migration Strategy:**

In real scenarios, users like to change cpu/memory limits online w/o restarting the service, show the current limit somewhere and receive alerting when the usage is approaching the limit.

### PR DESCRIPTION
**Pull request by pingandb**
### What problem does this PR solve?
There isn't yet a mechanism to monitor the cgroup cpu/memory limit and do corresponding alerting.

Issue Number: close #50467

Problem Summary:

### What changed and how does it work?
Cgmon pkg was added to periodically watch the cgroup cpu/memory limit setting changes. GOMAXPROCS will be set to the minor one between cgroup limit and maxprocs. Metrics will be set accordingly and displayed on Grafana. New alerting rules were also added for the cpu/memory usage ratio relative to the limits.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
        Grafana display and alerting trigger
        CPU/Memory Quota will be displayed with the usage graphs as another legend like below (playground on Mac m1, no real data):
image

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/util/cgmon/cgmon.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
+package cgmon
```
Comment (hawkingrei) [inline pkg/util/cgmon/cgmon.go:1]: need to add a license header.

you need to run ```make bazel_prepare``` to update the bazel config.

File: pkg/util/cgroup/cgroup_cpu_linux.go [hunk 2]
```diff
@@ -57,6 +57,11 @@ func CPUQuotaToGOMAXPROCS(minValue int) (int, CPUQuotaStatus, error) {
 	return maxProcs, CPUQuotaUsed, nil
 }
 
+// GetCPUPeriodAndQuota returns CPU period and quota time of cgroup.
+func GetCPUPeriodAndQuota() (int64, int64, error) {
```
Comment (hawkingrei) [inline pkg/util/cgroup/cgroup_cpu_linux.go:61]: ```getCgroupCPU``` has got the ```Period```, ```Quota``` of CPU.

Comment (pingandb) [reply pkg/util/cgroup/cgroup_cpu_linux.go:61]: > `getCgroupCPU` has got the `Period`, `Quota` of CPU.

Yes it does. But we only want to get period and quota here. Don't want to call getCgroupCPU to make additional calls to detectCPUUsageInV1/detectCPUUsageInV2.

**Comment by wuhuizuo** [comment]
@pingandb Thanks for your contribution, could you please rebase your branch.

File: pkg/util/cgmon/cgmon.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	if cpuPeriod > 0 && cpuQuota > 0 {
+		ratio := float64(cpuQuota) / float64(cpuPeriod)
+		if ratio < float64(quota) {
+			quota = int(math.Ceil(ratio))
+		}
+	}
+
+	if quota != lastMaxProcs && quota < cfgMaxProcs {
+		runtime.GOMAXPROCS(quota)
+		log.Info(fmt.Sprintf("maxprocs set to %v", quota))
```
Comment (hawkingrei) [inline pkg/util/cgmon/cgmon.go:116]:
```suggestion
		log.Info("set the maxprocs", zap.Int("quota",quota))
```

File: cmd/tidb-server/main.go [hunk 4]
```diff
@@ -372,8 +374,9 @@ func setCPUAffinity() {
 		fmt.Fprintf(os.Stderr, "set cpu affinity failure: %v", err)
 		os.Exit(1)
 	}
-	runtime.GOMAXPROCS(len(cpu))
-	metrics.MaxProcs.Set(float64(runtime.GOMAXPROCS(0)))
+	if len(cpu) < runtime.GOMAXPROCS(0) {
+		runtime.GOMAXPROCS(len(cpu))
```
Comment (hawkingrei) [inline cmd/tidb-server/main.go:378]: We can add log here.

Comment (pingandb) [reply cmd/tidb-server/main.go:379]: > We can add log here.

Done, thx.

**Review by hawkingrei — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Pull request merged**